### PR TITLE
[cni-cilium] Fix agent-wide deadlock and crash in ipcache/policy under identity churn.

### LINKED ISSUE #21647: cni-cilium: cilium-agent deadlock on cilium 1.17.x (ipcache/endpoint lock inversion), fixed only in cilium 1.18
**Deckhouse:** 1.75.14, `cni-cilium` (cilium **1.17.4**). Also applies to the 1.17.17 bump in #20720.

### Problem

`cilium-agent` deadlocks and never recovers on its own. Symptoms on the affected node:

* every endpoint stuck in `regenerating` / `waiting-to-regenerate`, none reach `ready`
* CNI returns `[PUT /endpoint/{id}][429] putEndpointIdTooManyRequests`, pod sandbox creation hangs 30–90 min
* agent RSS grows ~250–300 MiB/h (0.7 → 5.1 GiB observed), no OOM
* **liveness probe stays green, 0 restarts** — it does not touch the deadlocked locks
* the only recovery is deleting the agent pod

Seen twice in 10 days on a 60-node production cluster (3 nodes total), 15–19 h per occurrence, uncorrelated with any config change.

### Root cause (upstream, cilium 1.17.x)

Lock inversion between the IPCache and endpoint locks, line numbers for `v1.17.4`:

* `pkg/ipcache/metadata.go:541` — `doInjectLabels` takes `ipc.mutex.Lock()` (`defer Unlock()`)
* `:610` — still holding it, calls `UpdatePolicyMaps(ctx, nil, idsToDelete)`
* `:646` — `policyImplementedWG.Wait()`, no ctx/timeout: waits for **all** endpoints to apply their policy maps
* the other direction: endpoint regeneration holds the endpoint lock and needs the IPCache lock — upstream documents exactly this hazard in `pkg/endpoint/bpf.go:634` (*"Endpoint lock must not be held during 'GetDNSRules' as it locks IPCache, which leads to a deadlock"*)

The add path was deliberately moved out of the lock (`metadata.go:531`, comment mentions the deadlock); the delete path was not.

Related upstream report: cilium/cilium#40219 (same version, same 429, closed as stale).

### Fixed upstream in 1.18.0

Commit [`47ace2de6`](https://github.com/cilium/cilium/commit/47ace2de6) (PR cilium/cilium#39970, merged 2025-06-13) removes `UpdatePolicyMaps` from `doInjectLabels` and awaits `IdentityUpdater.UpdateIdentities()` **before** taking `ipc.mutex`. First tag: `v1.18.0-rc.0`, first stable: **`v1.18.0`**.

Not backported — `v1.17.17` and `v1.17.18` still carry the pattern (`:548` lock / `:617` `UpdatePolicyMaps` / `:653` `Wait`), so the 1.17.4 → 1.17.17 bump in #20720 does not address this.

### Ask

1. Ship cilium 1.18.x in `cni-cilium`, or backport `47ace2de6` to the vendored 1.17.
2. Until then: a health check that detects endpoints stuck in `regenerating` (e.g. `cilium_endpoint_state{endpoint_state=~"regenerating|waiting-to-regenerate"}` non-zero for >10 min) — the current liveness probe cannot see this state.

### PR DESCRIPTION
**Pull request by AndreyPavlovFlant**
## Description

Adds two backport patches for the Cilium 1.17.x agent used in the cni-cilium module:

- `019-ipcache-no-deadlock-on-label-injection.patch` — moves the delete-path UpdatePolicyMaps() call in doInjectLabels() out of the ipc.mutex critical section (matching the add path above it) and bounds the previously unbounded wait with ctx and a 3-minute timeout.
- `020-policy-nil-safe-selector-policy-detach.patch` — adds a nil check in policyCache.delete() before calling selectorPolicy.detach(), so a cache entry whose policy is still being resolved by updateSelectorPolicy() (without holding the cache lock) is not dereferenced.
Both patches document the exact upstream fix and are marked for removal on upgrade to Cilium 1.18+.

Fixes https://github.com/deckhouse/deckhouse/issues/21647

## Why do we need it, and what problem does it solve?

Under identity churn the Cilium agent could deadlock: doInjectLabels() held ipc.mutex while UpdatePolicyMaps() waited for all endpoints to regenerate, but endpoint regeneration acquires the same mutex (GetDNSRules → LookupByIdentity). As a result endpoint regeneration stopped node-wide, CNI requests piled up, the endpoint-create limiter started returning 429 putEndpointIdTooManyRequests, and pods failed to get networking — while agent liveness stayed green, so the agent was not restarted automatically.

Once regeneration ran at full speed, a second latent bug surfaced: policyCache.delete() dereferenced a nil SelectorPolicy for an endpoint whose identity changed mid-regeneration, crashing the agent with a SIGSEGV in selectorPolicy.detach.

Together these caused node-wide loss of pod networking and agent crashes on clusters with high identity churn.

## Why do we need it in the patch release (if we do)?

Both are production-impacting stability bugs (node-wide connectivity outage and agent crash) that affect existing clusters running the current Cilium 1.17.x. The fixes are self-contained backports and should be delivered to affected clusters without waiting for the Cilium upgrade, where both are fixed upstream.

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: cni-cilium
type: fix
summary: Fix agent-wide deadlock in ipcache label injection and a nil-pointer crash in policy cache under identity churn.
impact: Cilium agent pods are restarted to apply the fixes.
impact_level: high
```

### CODE REVIEW

**Review by apolovov — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Cherry pick PR [22586](https://github.com/deckhouse/deckhouse/pull/22586) to the branch [release-1.77](https://github.com/deckhouse/deckhouse/tree/release-1.77) successful!
